[ECO-5241] Ephemeral Typing Indicators

umair-ably opened the pull request:
Replaces presence based Typing Indicators with Ephemeral Typing Indicators as per [1].

Waiting for typing unit tests to be added again by Marat before extending them. Add unit tests as part of https://github.com/ably/ably-chat-swift/issues/256. The integration test has been extended to cover the heartbeat throttle. 

[1] - All commits up to and including https://github.com/ably/specification/pull/277/commits/9c1c8adb3c20816fd7be7f8ab773ae2e240cfbcf

---

Demo shows how typing within the 12 second period (10 sec heartbeat throttle + 2 sec timeout) does not impact the typing indicator. Subsequent typing restarts the typing indicator. Deleting the text stops the typing indicator.

https://github.com/user-attachments/assets/0a6126e4-e206-4c8a-82d1-76192f4d2c0a

---

Demo shows how we don't have hammering/duplicate event issues as outlined in the comments in `TypingTimerManager`. Repeatedly invoking `.keystroke` as shown in the video only has 1 corresponding event sent at the expected interval.

https://github.com/user-attachments/assets/0e604cb4-1379-4751-9d4c-855c07b7f123




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Typing indicators now use heartbeat throttling for efficient and accurate typing status updates.
  - Typing events include detailed change information for improved tracking.
- **Improvements**
  - Typing indicators are managed on the main chat channel, simplifying channel structure.
  - Enhanced timer and throttling logic prevents excessive typing event emissions.
  - Introduced serialized operation queueing for typing actions to ensure orderly event processing.
- **Bug Fixes**
  - Improved typing indicator timing and event handling for greater reliability.
- **Tests**
  - Added extensive tests covering typing behavior, timer management, and operation queue functionality.
- **Chores**
  - Updated dependencies and added new packages to support advanced timer and concurrency capabilities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lawrence-forooghian commented:
> Waiting for typing unit tests to be added again by Marat before extending them. 

@umair-ably we previously agreed that Marat wouldn't reintroduce the typing tests (see #235), so I would suggest starting the unit tests from scratch.

lawrence-forooghian commented:
Also can we please change this PR to target an integration branch for the single-channel work? per https://ably.atlassian.net/wiki/spaces/CHA/pages/3885826051/CHADR-097+API+Changes+Final+Steps+For+Single+Channel we want to coordinate the release of this work across all SDKs, and we don't want to be blocked from doing other releases of chat-swift in the meantime.

lawrence-forooghian commented:
@umair-ably I notice that there's been another update to the spec PR since the commit you reference here; please could you update here accordingly?

lawrence-forooghian commented on Sources/AblyChat/Typing.swift:
```diff
@@ -72,7 +73,23 @@ public struct TypingEvent: Sendable {
+    /**
+     * Get the details of the operation that modified the typing event.
+      */
+    public var change: Change
```
Where is this new public API taken from?

umair-ably replied:
added in V2 of the event https://sdk.ably.com/builds/ably/specification/pull/277/chat-features/#chat-structs-typing-event-V2

lawrence-forooghian commented on Sources/AblyChat/TypingTimerManager.swift:
```diff
@@ -0,0 +1,97 @@
+// (CHA-TM14b) Once the lock is acquired, if another call is made to either function, the second call shall be queued and wait until it can acquire the lock before executing.
+// (CHA-TM14b1) During this time, each new subsequent call to either function shall abort the previously queued call. In doing so, there shall only ever be one pending call and while the mutex is held, thus the most recent call shall “win” and execute once the mutex is released.
+
+/** The above spec points are handled implicitly because we use an actor `TypingTimerManager`, which in turn uses another actor `TimerManager`. This results in function calls to `TimerManager` needing to be awaited, which in turn require function calls from `TypingTimerManager` to be awaited. These awaited functions e.g. `startTypingTimer` ensure a "lock" is in place and subsequent calls are igno ...
```
I'm not sure that we're satisfying CHA-T14. I can't see a mechanism that does either of the following:

- prevents one call to, say `keystroke()` from completing whilst another is in-flight
- discards all requests except for the latest one

lawrence-forooghian replied:
https://github.com/ably/ably-chat-swift/issues/251 relevant here

umair-ably replied:
fixed

lawrence-forooghian commented on Sources/AblyChat/DefaultTyping.swift:
```diff
@@ -39,165 +46,142 @@ internal final class DefaultTyping: Typing {
-                    logger.log(message: "Failed to fetch presence set after \(maxRetryDuration) seconds. Giving up.", level: .error)
+
+                    // (CHA-T13) When a typing event (typing.start or typing.stop) is received from the realtime client, the Chat client shall emit appropriate events to the user.
+                    await subscription.emit(
```
Doesn't this emit another `started` event even if that user was already typing?

umair-ably replied:
fixed

lawrence-forooghian commented on Sources/AblyChat/DefaultTyping.swift:
```diff
@@ -39,165 +46,142 @@ internal final class DefaultTyping: Typing {
+                    await typingTimerManager.cancelTypingTimer(for: messageClientID)
+
+                    // (CHA-T13) When a typing event (typing.start or typing.stop) is received from the realtime client, the Chat client shall emit appropriate events to the user.
+                    await subscription.emit(
```
What about CHA-T13b5, which is meant to ignore the event if the user was not typing?

lawrence-forooghian commented on Sources/AblyChat/TypingTimerManager.swift:
```diff
@@ -0,0 +1,97 @@
+        return true
+    }
+
+    internal func startTypingTimer(for clientID: String, isSelf: Bool = false, handler: (@Sendable () -> Void)? = nil) async {
```
The spec defines two different kinds of timer, which:

- have different durations
- are meant to be started in response to different events
- have different behaviours when they fire

These are:

- the "I am typing" timer of CHA-T4a4
- the "someone is typing" timers of CHA-13b1 (one of these timers might be for our client ID, but that doesn't mean that we can treat this timer the same as the CHA-T4a4 one)

`TypingTimerManager` does not make this distinction between types of timer, which is going to lead to incorrect behaviour.

umair-ably replied:
fixed

lawrence-forooghian replied:
I think that these two things are still kind of muddled (perhaps trying to manage them both in the "timer manager" class isn't helping) — let's discuss

lawrence-forooghian commented on Sources/AblyChat/TypingTimerManager.swift:
```diff
@@ -0,0 +1,97 @@
+    // Stores only the current user's last typing time
+    private var lastTypingTime: Date?
+
+    internal init(heartbeatThrottle: TimeInterval = 10, timeout: TimeInterval = 2, logger: InternalLogger) {
```
(Don't worry about sorting it out in `Room.swift` though; I've got that covered in the single-channel stuff)

umair-ably replied:
sorted

umair-ably commented on Sources/AblyChat/TypingTimerManager.swift:
```diff
@@ -0,0 +1,97 @@
+/** The above spec points are handled implicitly because we use an actor `TypingTimerManager`, which in turn uses another actor `TimerManager`. This results in function calls to `TimerManager` needing to be awaited, which in turn require function calls from `TypingTimerManager` to be awaited. These awaited functions e.g. `startTypingTimer` ensure a "lock" is in place and subsequent calls are igno ...
+internal final actor TypingTimerManager {
+    private let heartbeatThrottle: TimeInterval
+    private let timeout: TimeInterval
```
good shout, sorted

umair-ably commented on Sources/AblyChat/DefaultTyping.swift:
removed

umair-ably commented:
> @umair-ably I notice that there's been another update to the spec PR since the commit you reference here; please could you update here accordingly?

reviewed and updated. also updated the commit message

umair-ably commented:
> Also can we please change this PR to target an integration branch for the single-channel work? per https://ably.atlassian.net/wiki/spaces/CHA/pages/3885826051/CHADR-097+API+Changes+Final+Steps+For+Single+Channel we want to coordinate the release of this work across all SDKs, and we don't want to be blocked from doing other releases of chat-swift in the meantime.

we seem to have just gone straight to main in kotlin? do we have an integration branch already setup?

umair-ably commented:
ignore that i've directly copied the AsyncSemaphore class instead of adding the package... just fiddling with why the CI could be failing

lawrence-forooghian commented:
> we seem to have just gone straight to main in kotlin? 

Can't speak for what was done in Kotlin, @ttypic why did you folks decide not to do an integration branch for single-channel (see https://github.com/ably/ably-chat-swift/pull/250#issuecomment-2766119331)?

> do we have an integration branch already setup?

No, can you make one please?

lawrence-forooghian commented on Sources/AblyChat/AsyncSemaphore.swift:
@umair-ably why did you decide to include this directly? We haven't given any attribution to the original author.

umair-ably replied:
i left a comment... was just messing around with why it could be failing https://github.com/ably/ably-chat-swift/pull/250#issuecomment-2775377627... i've reverted to using the package

lawrence-forooghian commented:
@umair-ably what's your plan for adding tests for this functionality?

ttypic commented:
Yeah, we can discuss this, but I don’t like integration branches. The main reason is that we spend most of our time working on new features, which, in this case, live in the integration branch. This branch essentially becomes our main branch—we collaborate on it and do all our work there. But if that’s the case, why not just work directly on the actual main branch?

You might argue that we still need to add fixes to older releases, but in my experience, that’s rare. When it happens, it’s better to create a fix branch from the previous tag and push the changes from there. This might slightly conflict with the collaboration guide, but it’s still straightforward and, in my opinion, infrequent enough to be manageable.

umair-ably commented:
> > @umair-ably what's your plan for adding tests for this functionality?
> 
> I'll add them when I'm back next week... Marat's other tests/mocks should be in by then too... i'll create an issue and link it here before this gets merged

https://github.com/ably/ably-chat-swift/issues/256

lawrence-forooghian commented:
Have rebased this on top of #261. One of the `TaskQueue` tests is failing; will look into it.

ttypic commented on Package.swift:
```diff
@@ -38,6 +38,10 @@ let package = Package(
         ),
+        .package(
+            url: "https://github.com/groue/Semaphore",
+            from: "0.1.0"
```
- do you think it’s safe to use version 0.1.0? Could it potentially cause dependency issues? Would it be better to copy-paste the code instead? (It’s just 200 lines of code in a single file.) In jvm/js/python, I definitely wouldn’t do this, but I’m not sure how dependency resolution works in Swift.
- do you think using a semaphore is fair? Looking at the source code, it seems to use NSRecursiveLock, which isn’t fair

umair-ably replied:
Given the repo is just 1 release from July 2024, which is also a very simple file, I'd say it's improbable this will cause us an issue.

I can lock it to exactly `0.1.0` if that makes us feel better but I'd rather trust any breaking changes will follow semantic versioning rules.  The `from` parameter will prevent an update to a major release

lawrence-forooghian replied:
> Would it be better to copy-paste the code instead?

Umair had done this in an earlier version of this PR but I got a bit worried re licensing — we'd have to make sure we properly attribute it in a way that respects the licensce

ttypic replied:
I think you can do whatever you want if source code is under MIT. The MIT license permits you "to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software"

lawrence-forooghian commented:
@umair-ably please can you fix the `spec-coverage` job so that I can see the spec coverage for this feature? I think you have two things to fix:

1. It thinks that your "explanation of what the spec point is" comments are part of the spec point name. If you want to include such comments I'd suggest including them on the line above; the `@spec…` tag structure permits you to add a comment, but said comment is meant to explain something about the test itself
2. It's also going to fail since you refer to spec points that aren't on the spec `main` branch; you need to add a temporary commit similar to https://github.com/ably/ably-chat-swift/pull/247/commits/d6b2a716f60094527e2a3c9ead2f0cd367020ca7 to run it against the revision of the spec that you're targeting

lawrence-forooghian commented on Sources/AblyChat/DefaultTyping.swift:
```diff
@@ -35,229 +57,194 @@ internal final class DefaultTyping: Typing {
-                } else {
-                    // (CHA-T5a) If typing is not in progress, this operation is no-op.
-                    logger.log(message: "User is not typing. No need to leave presence.", level: .debug)
+                    if await typingTimerManager.isHeartbeatTimerActive() {
```
Are you sure?

lawrence-forooghian commented on Sources/AblyChat/DefaultTyping.swift:
```diff
@@ -35,229 +57,194 @@ internal final class DefaultTyping: Typing {
-                                continuation.resume(returning: .failure(error))
-                            }
-                        }
+                    guard await !typingTimerManager.isHeartbeatTimerActive() else {
```
Wouldn't the compiler have complained?

lawrence-forooghian commented on Sources/AblyChat/DefaultTyping.swift:
```diff
@@ -4,205 +4,239 @@ internal final class DefaultTyping: Typing {
-                            logger.log(message: "User is not typing. Starting typing.", level: .debug)
-                            do throws(InternalError) {
-                                try startTyping()
-                                continuation.resume(returning: .success(()))
-                            } catch {
-                                continuation.resume(returning: .failure(error))
-                            }
-                        }
+                    guard !typingTimerManager.isHeartbeatTimerActive else {
+                        // (CHA-T4c) If typing is already in progress (i.e. a heartbeat timer set according to CHA-T4a4 exists and has not expired):
+                        // (CHA-T4c1) The client must not send a typing.started event.
+                        logger.log(message: "Throttle time hasn't passed, skipping typing event.", level: .debug)
+                        return
                     }
-                }.get()
+
+                    try await publishStartedEvent()
+                }
             } catch {
                 throw error.toARTErrorInfo()
             }
         }
```
> Swift doesn't support specifying the error type using throws(InternalError) syntax.

Yes it does, please remember this

lawrence-forooghian commented on Tests/AblyChatTests/TypingOperationQueueTests.swift:
```diff
@@ -0,0 +1,174 @@
+
+    @Test
+    func enqueue_rethrowsOperationError_whenOperationExecutedFromIdle() async {
+        let queue = TypingOperationQueue<TestError>()
+
+        await #expect(throws: TestError.badThing) {
+            try await queue.enqueue { () async throws(TestError) in
+                throw TestError.badThing
+            }
+        }
+    }
```
This syntax is supported, plesae remember this

lawrence-forooghian commented:
@coderabbitai stop

lawrence-forooghian commented on Tests/AblyChatTests/IntegrationTests.swift:
```diff
@@ -335,13 +335,14 @@ struct IntegrationTests {
             typingEvents.append(typingEvent)
             if typingEvents.count == 1 { break }
+            break
         }
```
Have removed the subsequent `break`, @umair-ably when back please check this matches your intention

lawrence-forooghian commented:
Something in one of `DefaultTypingTests` or `TypingTimerManagerTests` is causing CI to hang, and I haven't yet figured out what. I need to get the bulk of the typing indicators work merged, so that I can merge my single-channel work in #247 (else I think people will struggle with the rebase and updating commit messages if they try to do it whilst I'm holiday). So, I've added a commit which deletes these tests. Have created #283 for us to fix and restore them.

lawrence-forooghian reviewed (approved):
Happy with the implementation, haven't looked at the tests but will do so when they're reintroduced in #283.
